### PR TITLE
Navigation

### PR DESCRIPTION
**Pull request by nilslindemann**
Like promised [here](https://github.com/mint-lang/mint/discussions/614), I took a look at how navigation works in Mint, and I (hopefully) streamlined things a bit. It is best explained by example, so open

[mint-navi-test.tar.gz](https://github.com/mint-lang/mint-runtime/files/11751501/mint-navi-test.tar.gz)

and compare `dist now (current issues)` with `dist then 1 (issues resolved)`. Run e.g. `npx serve` in the folders, or [use the tool of your choice](https://gist.github.com/willurd/5720255). You may have to reload without cache, <kbd>ctrl</kbd>+<kbd>f5</kbd>.

If you want to check my changes using the mint development server,

* Apply [the PR in the Mint repo](https://github.com/mint-lang/mint/pull/616).
* Apply this PR here.
* Run `yarn install` here. (there is a new dependency)
* Apply the three changes in [the PR in the route-parser repo](https://github.com/rcs/route-parser/pull/39) directly to `node_modules/route-parser/lib/route/visitors/regexp.js` here.
* Run `make` here.
* Run `make development` in the mint repo.
* Run `mint-dev start` in the `mint-navi-test` folder from above.

(make sure these three folders are side by side)

None of these changes are applied in `dist now` from the above zip. `dist then 1` and `dist then 2` have all of them applied. `dist then 2` shows the differences between `Window.setUrl()`,  `Window.navigate()` and  `Window.jump()`.

So, this is my suggestion how we could do the navigation.

### CODE REVIEW

**Comment by nilslindemann** [comment]
Ok, I tried running that linter locally (`yarn lint`), but I get the same error message. And the SO answer suggesting fixes lead to other errors, like weird network connections. I resign on that, Linter wins :-)

**Comment by gdotdesign** [comment]
> Ok, I tried running that linter locally (`yarn lint`), but I get the same error message. And the SO answers suggesting fixes lead to other errors, like weird network connections. I resign on that, linter wins :-)

You can format the files with `yarn format`

**Comment by gdotdesign** [comment]
I've tried `dist now` and `dist then 2` and can't seem to find differences in how they work. Can you summary what the differences are please?

File: source/Program.js [hunk 1]
```diff
@@ -89,83 +106,101 @@ export default (enums) => {
...
+              throw new DecodingError();
+            }
+          });
+          route.handler.apply(null, args);
+        } catch (error) {
+          if (error.constructor !== DecodingError) {
+            throw error;
           }
-        }
+        } finally {}
```
[the remark this answers is not in the document]
Comment (nilslindemann) [reply source/Program.js:202]: We do not. It was in the original code, I do not remember why I did not remove it. Did now.

**Comment by nilslindemann** [comment]
> You can format the files with `yarn format`

Thank you.

File: source/Program.js [hunk 2]
```diff
@@ -89,83 +106,101 @@ export default (enums) => {
...
+                this.root.querySelector(hash)
+                ||
+                this.root.querySelector(`a[name="${hash.slice(1)}"]`)
+                ;
+            } finally {}
+
+            if (elem) {
+              if (triggerJump) {
+                if (isSameRoute) {
+                  elem.scrollIntoView({ behavior: "smooth" });
```
[the remark this answers is not in the document]
Comment (nilslindemann) [reply source/Program.js]: Ok, I removed that code. I actually was not sure myself if it is a good idea.

**Comment by nilslindemann** [comment]
> I've tried `dist now` and `dist then 2` and can't seem to find differences in how they work. Can you summary what the differences are please?

Ignore `dist then 2`. I just included that because I tested with it most of the time. It is more confusing than helpful, I guess.

I have recreated [mint-navi-test.tar.gz](https://github.com/mint-lang/mint-runtime/files/12211386/mint-navi-test.tar.gz) with hopefully better explanation. Run `npx serve` in `dist now` and then run it in `dist then`, as described in my first post. (Do not forget to press <kbd>ctrl</kbd>+<kbd>f5</kbd>, the Node.js server or the browser anyhow caches things).

* `dist now` is based on the main branch, speak that is how it currently behaves.
* `dist then` is all these commits and other changes applied, including the changes in the [mint PR](https://github.com/mint-lang/mint/pull/616) and in the [route-parser PR](https://github.com/rcs/route-parser/pull/39). I show and explain the changes there.

**Comment by nilslindemann** [comment]
And here is some manual explanation from memory, what this and the other PRs do:

* When clicking links, if the link does not contain a hash, it now jumps to the top of the document.
* history forward / back now works as expected.
* triggering route logic is more intuitive. A route `/route` will be triggered by `/route`, `/route#hash`, `/route?seach` and `/route?search#hash`.
* However, only if the path part was not already `/route`.
* One can also define a more specific route like `/route#foo` which will then catch this specific hash, but e.g. `\route#bar` will still be catched by the generic `/route`.
* Variables can be used everywhere, e.g. `/route/:p?query=:q#:h`.
* A '#' will not be consumed by a variable anymore. So the hash will not be accidently consumed. So, e.g. the route `/:p#hash` will actually match e.g. `/foo#hash` and `p` will be `foo`. (and not `foo#hash` like before, which resulted in the route not matching)
* Additionally to the Mint functions `Window.setUrl()` and `Window.navigate()` there is now also a function `Window.jump()`. `setUrl()` behaves like before, it just sets the URL in the browser address bar. `navigate()` additionally updates the history and maybe triggers the route logic. `jump()` additionally jumps to the hash or start of document.

**Comment by gdotdesign** [comment]
Thanks for the explanation, it's clearer to me now (also I had to hard hard refresh Shift+Ctrl :sweat_smile: )

I agree with most of the changes, however not with this:

> However, only if the path part was not already /route

I usually have the search value as a query parameter `?search=xxx` and when that parameter changes I reload the content matching that search and currently it means that the route will be triggered. I do this because I believe the URL should be the source of truth of the UI (everything can be derived from it). Also, in this case I replace the URL instead of pushing a new one, so the history remains correct.

The second use case is hash based navigation `/#/route`, which would not work with this change (it doesn't work now because of the bug you discovered).

**Comment by nilslindemann** [comment]
@gdotdesign 

I forgot to mention this: A route gets (re-)triggered in three cases:

* first load of the app
* the path part changed
* an argument for a parameter changed.

This is done [here in the code](https://github.com/mint-lang/mint-runtime/blob/6fa42f1e670a75c1c1c4f94f8bac000980af669a/source/Program.js#L163-L165). (I call the parameters `vars` there)

So, If one defines a route `/foobar?search=:xxx` then it will get triggered, whenever the value for the `xxx` changes. If one only defines a route `/foobar` then it will not get triggered again when something in the search (or hash) part changes.

It is also possible to define a route `foobar/#:hash`, and it will get triggered again, whenever the argument for `:hash` changes.

It is correct, a variable can not catch a `/`  see [this regex](https://github.com/rcs/route-parser/blob/903d07a62fe7649fccb6be6694445ee4217c933e/lib/route/visitors/regexp.js#L74). But my PR actually did not change that, I just added the '#' to the regex. We may change it to allow `.*`, but I have not tested yet if that works.

**Review by Sija — APPROVED** [review]
LGTM 🚀

**Comment by nilslindemann** [comment]
There is one issue: Currently, no one is reacting to the PR in the [route-parser](https://github.com/rcs/route-parser/pull/39) repo. So these changes will not be applied when one does `yarn install` here. The simplest solution would probably be to fork that repo as e.g. `mint-route-parser`, apply the PR and use that forked repo instead as [dependency](https://stackoverflow.com/questions/36078409/git-repository-to-a-dependency-in-package-json) in [package.json](https://github.com/mint-lang/mint-runtime/blob/d392ba41f2942dd874c793f0c31c401f768f8f4e/package.json#L28). There is also something like [patch-package](https://www.npmjs.com/package/patch-package#why-use-postinstall-postinstall-with-yarn), but I have not tested that.

The other questions, should we add a fourth function `triggerHandler`, should we merge these four functions into one if Mint gets named arguments, can be answered later.

**Comment by Sija** [comment]
I'd take over the library or use `patch-package`.

**Comment by nilslindemann** [comment]
I have just tried replacing the dependency (`^0.0.5`) with `nilslindemann/route-parser#let-hash-not-be-part-of-parameters`, which is the above PR. Then deleted `node_modules` folder and ran `yarn install`. This worked. (using yarn version 1.22.19)

So something like `mint-lang/mint-route-parser#main` should probably work.

**Comment by nilslindemann** [comment]
Great stuff :pray:

**Pull request merged**